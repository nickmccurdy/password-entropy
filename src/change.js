export default function (wrapper, target) {
  wrapper.simulate('change', { target: { ...wrapper.props(), ...target } })
}
