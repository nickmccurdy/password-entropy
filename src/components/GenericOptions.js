import handleChange from '../handleChange'
import Options from './Options'
import { filter, identity, keys, map, pipe, sum, T } from 'ramda'
import React, { PureComponent } from 'react'
import { Checkbox } from 'react-bootstrap'

export default class GenericOptions extends PureComponent {
  static propTypes = Options.sharedPropTypes
  static shortName = 'Generic'

  static toggles = {
    letters: {
      label: 'Letters',
      example: 'a-z',
      possibleItems: 26
    },
    capitalLetters: {
      label: 'Capital Letters',
      example: 'A-Z',
      possibleItems: 26
    },
    numbers: {
      label: 'Numbers',
      example: '0-9',
      possibleItems: 10
    },
    symbols: {
      label: 'Symbols',
      example: '!@#$%^&*',
      possibleItems: 8
    }
  }

  state = map(T, this.constructor.toggles)

  handleChange = handleChange.bind(this)

  possiblePasswords = () => pipe(filter(identity),
                                 keys,
                                 map(key => this.constructor.toggles[key].possibleItems),
                                 sum)(this.state)

  render () {
    return (
      <Options possiblePasswords={this.possiblePasswords} {...this.props}>
        <h3>Generic</h3>

        {Object.entries(this.constructor.toggles)
               .map(([name, { label, example }]) => (
                 <Checkbox key={name} name={name} checked={this.state[name]} onChange={this.handleChange}>
                   { label } <small>({ example })</small>
                 </Checkbox>
               ))}
      </Options>
    )
  }
}
