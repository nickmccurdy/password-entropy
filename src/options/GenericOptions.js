import startCase from 'lodash/startCase'
import PropTypes from 'prop-types'
import always from 'ramda/src/always'
import map from 'ramda/src/map'
import path from 'ramda/src/path'
import T from 'ramda/src/T'
import React from 'react'
import Checkbox from 'react-bootstrap/lib/Checkbox'
import { connect } from 'react-redux'

import { toggleGeneric } from '../actions'

export const toggles = {
  letters: {
    example: 'a-z',
    possibleItems: 26
  },
  capitalLetters: {
    example: 'A-Z',
    possibleItems: 26
  },
  numbers: {
    example: '0-9',
    possibleItems: 10
  },
  symbols: {
    example: '!@#$%^&*',
    possibleItems: 8
  }
}

export const defaultToggles = map(T, toggles)

export function GenericOptions(props) {
  return (
    <div>
      {Object.entries(toggles).map(([name, { example }]) => (
        <Checkbox key={name} name={name} checked={props[name]} onChange={props.toggleGeneric}>
          { startCase(name) } <small>({ example })</small>
        </Checkbox>
      ))}
    </div>
  )
}

GenericOptions.propTypes = {
  ...map(always(PropTypes.bool.isRequired), toggles),
  toggleGeneric: PropTypes.func.isRequired
}

export const mapStateToProps = path(['options', 'generic'])

export default connect(mapStateToProps, { toggleGeneric })(GenericOptions)
