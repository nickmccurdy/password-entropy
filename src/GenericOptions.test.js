import change from './change'
import { mount, shallow } from 'enzyme'
import GenericOptions from './GenericOptions'
import React from 'react'

const callback = jest.fn()
const node = (<GenericOptions onChange={callback}/>)

beforeEach(callback.mockReset)

it('renders', () => {
  expect(shallow(node)).toMatchSnapshot()
})

it('calls onChange prop', () => {
  const wrapper = mount(node)
  expect(callback).toHaveBeenCalledWith(70)
  expect(wrapper.state()).toEqual({
    "letters": true,
    "capitalLetters": true,
    "numbers": true,
    "symbols": true
  })

  callback.mockReset()
  change(wrapper.find({ name: 'letters' }), { checked: false })
  expect(callback).toHaveBeenCalledWith(44)
  expect(wrapper.state()).toEqual({
    "letters": false,
    "capitalLetters": true,
    "numbers": true,
    "symbols": true
  })
})
