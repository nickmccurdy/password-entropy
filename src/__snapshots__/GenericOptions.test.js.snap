// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<Options
  onChange={[Function]}
  possiblePasswords={[Function]}
>
  <h3>
    Generic
  </h3>
  <label>
    <input
      checked={true}
      name="letters"
      onChange={[Function]}
      type="checkbox"
    />
    Letters
  </label>
  <label>
    <input
      checked={true}
      name="capitalLetters"
      onChange={[Function]}
      type="checkbox"
    />
    Capital Letters
  </label>
  <label>
    <input
      checked={true}
      name="numbers"
      onChange={[Function]}
      type="checkbox"
    />
    Numbers
  </label>
  <label>
    <input
      checked={true}
      name="symbols"
      onChange={[Function]}
      type="checkbox"
    />
    Symbols
  </label>
</Options>
`;
